src/transaction/transaction.rs

```diff
@@ -173,31 +168,130 @@ impl Transaction {
     /// # use tikv_client::{Config, transaction::Client};
     /// # use futures::prelude::*;
     /// # futures::executor::block_on(async {
-    /// # let connect = Client::connect(Config::default());
-    /// # let connected_client = connect.await.unwrap();
-    /// let mut txn = connected_client.begin().await.unwrap();
+    /// # let client = Client::new(Config::default()).await.unwrap();
+    /// let mut txn = client.begin().await.unwrap();
     /// // ... Do some actions.
     /// let req = txn.commit();
     /// let result: () = req.await.unwrap();
     /// # });
     /// ```
     pub async fn commit(&mut self) -> Result<()> {
+        TwoPhaseCommitter::new(
+            self.buffer.to_proto_mutations(),
+            self.timestamp.into_version(),
+            self.bg_worker.clone(),
+            self.rpc.clone(),
+        )
+        .commit()
+        .await
+    }
+}
+
+/// The default TTL of a lock in milliseconds
+const DEFAULT_LOCK_TTL: u64 = 3000;
+
+#[derive(new)]
+struct TwoPhaseCommitter {
+    mutations: Vec<kvrpcpb::Mutation>,
+    start_version: u64,
+    bg_worker: ThreadPool,
+    rpc: Arc<PdRpcClient>,
+    #[new(default)]
+    committed: bool,
+    #[new(default)]
+    undetermined: bool,
+}
+
+impl TwoPhaseCommitter {
+    async fn commit(mut self) -> Result<()> {
+        if self.mutations.is_empty() {
```

Comment:
I think we should check if `mutations` is read only, i.e., we can return here if the only values in mutations are cached values.

Reply:
If the mutations are empty, means the transaction is  read-only

Reply:
We are storing some cached read values in the mutations table too, so even if it is non-empty, it might still be a read-only transaction

Reply:
The buffer may contain cached read values. But when the buffer is transformed to the `mutations` here, it should contain mutations only.

Reply:
Oh yeah, so it does. I added a comment there to hopefully make that clearer.